crates/agglayer-storage/src/tests/db_generator.rs

```diff
@@ -0,0 +1,773 @@
+                .entries_per_cf
+                .entry("latest_pending_certificate_per_network_cf".to_string())
+                .or_insert(0) += 1;
+        }
+
+        // 3. LatestProvenCertificatePerNetwork (use one of the state certificates)
+        if config.certificates_per_network > 1 {
+            let proven_height = Height::from(config.certificates_per_network - 2);
+            let certificate = state_result
+                .certificates
+                .get(&(*network_id, proven_height.as_u64()))
+                .expect("Certificate should exist from state generation")
+                .clone();
+            let cert_id = certificate.hash();
+            let proven_cert = ProvenCertificate(cert_id, *network_id, proven_height);
+            db.put::<LatestProvenCertificatePerNetworkColumn>(network_id, &proven_cert)?;
+            *result
+                .entries_per_cf
+                .entry("latest_proven_certificate_per_network_cf".to_string())
+                .or_insert(0) += 1;
+        }
+    }
+
+    // 4. ProofPerCertificate: Generate proofs if requested
+    if config.generate_proofs {
+        // Generate a dummy proof for the first certificate
+        if let Some(cert_id) = state_result.certificate_ids.first() {
+            let proof = Proof::dummy();
+            db.put::<ProofPerCertificateColumn>(cert_id, &proof)?;
+            *result
+                .entries_per_cf
+                .entry("proof_per_certificate_cf".to_string())
+                .or_insert(0) += 1;
+        }
+    }
+
+    Ok(result)
+}
+
+/// Generate epochs database using the same certificates from generate_state_db
+pub fn generate_epochs_db(
+    path: &Path,
+    config: &GeneratorConfig,
+    state_result: &GenerationResult,
+) -> Result<GenerationResult, crate::storage::DBError> {
+    let db = DB::open_cf(path, epochs_db_cf_definitions())?;
+    let mut rng = rand::rngs::StdRng::seed_from_u64(config.seed);
+
+    let mut result = GenerationResult {
+        entries_per_cf: std::collections::HashMap::new(),
+        network_ids: state_result.network_ids.clone(),
+        certificate_ids: Vec::new(),
+        certificates: std::collections::HashMap::new(),
+    };
+
+    // Calculate how many complete epochs we'll have based on
+    // certificates_per_network
+    let num_epochs = config
+        .certificates_per_network
+        .div_ceil(CERTIFICATES_PER_EPOCH);
+
+    // Generate data for each network
+    for network_id in &state_result.network_ids {
+        // Generate certificates grouped by epochs
+        for epoch_idx in 0..num_epochs {
+            // Calculate how many certificates are in this epoch
+            let certs_in_this_epoch =
+                if (epoch_idx + 1) * CERTIFICATES_PER_EPOCH <= config.certificates_per_network {
+                    CERTIFICATES_PER_EPOCH
+                } else {
+                    config.certificates_per_network - (epoch_idx * CERTIFICATES_PER_EPOCH)
+                };
```

Comment:
I am not sure what is this calculation supposed to do. Is it something like this?

```suggestion
            let certs_in_this_epoch = std::cmp::min(
                config.certificates_per_network.saturating_sub(epoch_idx * CERTIFICATES_PER_EPOCH),
                CERTIFICATES_PER_EPOCH,
            );
```

But maybe it would be easier to loop over height and calculate epoch and cert index from it using division and modulus, idk.